src/python/torchdistx/_C/fake.cc

```diff
@@ -6,15 +6,40 @@
 
 #include "module.h"
 
+#include <ATen/Context.h>
 #include <ATen/Tensor.h>
+#include <torch/csrc/utils/cuda_lazy_init.h>
 #include <torch/csrc/utils/pybind.h>
 #include <torchdistx/fake.h>
 
 namespace torchdistx::python {
+namespace {
+
+void pyEnterFakeMode(bool fake_cuda) {
+  enterFakeMode(fake_cuda);
+
+  // If CUDA is not available, suppress PyTorch's attempt to initialize its CUDA
+  // subsystem which would fail and prevent us from instantiating CUDA devices.
+  if (fake_cuda) {
+    if (!at::hasCUDA()) {
+      torch::utils::set_requires_cuda_init(false);
+    }
+  }
+}
+
+void pyLeaveFakeMode() {
+  leaveFakeMode();
+
+  if (!isFakeModeActive() && !at::hasCUDA()) {
```

Comment:
might be missing something here - if ATen hasCuda() is false, why do we want to set_requires_cuda to true here?

Reply:
It is necessary so that once fake mode is over, the default behavior is restored. We tell PyTorch that it should attempt to initialize CUDA if requested by the user (which was the state before entering fake mode). Since we don't have CUDA available, that attempt will fail with an expected error (e.g. "PyTorch is not built with CUDA" or "No CUDA runtime is available").